rust/chains/hyperlane-ethereum/src/aggregation_ism.rs

```diff
@@ -0,0 +1,117 @@
+#![allow(clippy::enum_variant_names)]
+#![allow(missing_docs)]
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use ethers::providers::Middleware;
+use tracing::instrument;
+
+use hyperlane_core::{
+    AggregationIsm, ChainResult, ContractLocator, HyperlaneAbi, HyperlaneChain, HyperlaneContract,
+    HyperlaneDomain, HyperlaneMessage, HyperlaneProvider, RawHyperlaneMessage, H256,
+};
+
+use crate::contracts::i_aggregation_ism::{
+    IAggregationIsm as EthereumAggregationIsmInternal, IAGGREGATIONISM_ABI,
+};
+use crate::trait_builder::BuildableWithProvider;
+use crate::EthereumProvider;
+
+pub struct AggregationIsmBuilder {}
+
+#[async_trait]
+impl BuildableWithProvider for AggregationIsmBuilder {
+    type Output = Box<dyn AggregationIsm>;
+
+    async fn build_with_provider<M: Middleware + 'static>(
+        &self,
+        provider: M,
+        locator: &ContractLocator,
+    ) -> Self::Output {
+        Box::new(EthereumAggregationIsm::new(Arc::new(provider), locator))
+    }
+}
+
+/// A reference to an AggregationIsm contract on some Ethereum chain
+#[derive(Debug)]
+pub struct EthereumAggregationIsm<M>
+where
+    M: Middleware,
+{
+    contract: Arc<EthereumAggregationIsmInternal<M>>,
+    domain: HyperlaneDomain,
+}
+
+impl<M> EthereumAggregationIsm<M>
+where
+    M: Middleware + 'static,
+{
+    /// Create a reference to a mailbox at a specific Ethereum address on some
+    /// chain
```

Comment:
im surprised we need all this

Reply:
All ISMs have this boilerplate. It's required for calling `self.build_ethereum(conf, &locator, metrics, h_eth::AggregationIsmBuilder {})`, which has a trait bound `B: BuildableWithProvider + Sync`